### PR TITLE
fix(codex): 熔断切模型后的上下文压缩风暴

### PR DESCRIPTION
**Pull request by zqchris**
### 摘要

Codex 会话中途切模型后，上游继续按**切换前**那个模型的窗口判定超窗，导致无限压缩循环：压缩 → 重建后仍是同样大小 → 再压缩。本 PR 在 Cindy 侧熔断这个循环，并让故障对用户可见。

实测证据（2026-08-04，session `71a9b69a`，rollout `019fcd52`）：
- 15:09 起会话 model=`codex/gpt-5.6-sol`，`model_context_window`=258400。
- 15:19 第 1 次压缩**正常生效**：last input 241972 → 30738（压缩机制本身没坏）。
- 15:23 用户切模型到 `claude-opus-5`（`thread_settings_applied`），随后注入 21391 字符的 `<model_switch>` developer 消息。
- 15:24 起连续 8 次压缩，每次压缩后的 input 恒定在 325868 / 325922 / 327909 / 326767 / 326827 / 326952 / 327383 —— 彼此相差不到 0.5%。
- 全程 `model_context_window` **始终是 258400**（旧模型的值），从未更新。

关键在于**这种状态下压缩不可能收敛**：每次压完 `replacement_history` 只剩 9 条消息 / 44136 字符（第 5 次甚至只剩 4 条 / 7481 字符），膨胀的 ~315k 全在 system prompt 与 16 个 MCP provider 的工具定义里，压缩机制根本碰不到那部分。26 分钟内 9 次压缩，直到用户手动切走 harness 才停。

**窗口核算与压缩决策都在上游 codex-rs，本仓改不了**（`packages/maker-core/src/agents/codex/` 只是经 app-server 转发 `thread/settings/update`；Cindy 的 `AutoCompactController` 只服务 claude-code agent，codex 路径上只有 `memoryFlushController` 在观测）。这里只做本地能做的两件事：

1. **压缩风暴熔断**（新增 `compaction-storm.ts`）。连续 3 次「压了等于没压」即熔断，走与既有 `retry-escalation.ts` 完全相同的收口路径（终态错误 + turn 收口 + `turn/interrupt`），并给出可操作诊断（切回原模型 / 新开任务）。判据**只看压缩自身效果，不看窗口值** —— 上游窗口本身就是陈旧的，而我们核实过的窗口又是新模型的真实值，按任一个算都不会触发；「压了等于没压」是这个故障唯一稳定可观测的特征。
   顺带把 `onCompactBoundary` 改为**无条件注册**：此前它只在 Maker Memory 开启时才挂上去，关掉 Maker Memory 的用户连压缩边界都收不到，熔断永远不会触发。

2. **`capContextWindow` 不再取 `min(verified, reported)`**。取小的前提是「上报值可信、只是可能虚高」，而切模型后的陈旧上报值恰好**更小**，`min()` 会把已经算对的新模型窗口又拉回旧值 —— 上下文占比与 memory flush 阈值全程按错的窗口走。已核实窗口本身就是该路由的上限，直接用它不会虚高。

### 变更类型

- [ ] `feat` 新功能
- [x] `fix` 缺陷修复
- [ ] `refactor` / `perf` 重构或性能优化
- [ ] `docs` / `test` / `chore` 文档、测试或工程维护
- [ ] 其他：

### 范围

- 关联 Issue / 需求：无（用户实测故障；上游部分另行提 codex-rs issue，材料见下方「上游责任边界」）
- 本 PR 包含：codex 上下文压缩不收敛的本地熔断与诊断（含它所依赖的窗口核算修正）
- 明确不包含：
  - 不修上游 codex-rs 的 `model_context_window` 核算 —— 那是 codex 内部压缩决策，本仓碰不到
  - 不实现 Cindy 侧的 codex auto-compact（codex 自己有，不重造）
  - 不动 claude-code / pi agent 的窗口与压缩路径
  - 不处理「system prompt + MCP 工具定义本身过大」这个独立问题
- 用户可见变化：压缩不收敛时会看到一条终态错误（四语文案，新增 `logic.errors.codexCompactionNotConverging`），而不是无限循环卡住；此前这个故障对用户完全不可见。
- 是否存在 breaking change：无

### UI 变化

新增一条终态错误提示文案（四语），复用既有 ErrorBanner / ErrorMessageCard 渲染，无新组件、无布局、无样式改动。

- 引用的设计规范：**DESIGN.md §11.1 Voice & Content** —— 「Errors = what happened + what to do」：文案先说发生了什么（Codex 反复压缩上下文却压不动，本轮已自动中断以免无限循环），再给可执行的下一步（切回原模型继续，或新开一个任务），不是裸的「出错了」。**§11.3 Self-Check** 逐条核对：无「成功 / successfully」filler；四语 `common.json` 全部更新，各按本语言的大小写与标点约定（en 正文用 sentence case，zh-CN 不用英式标点，ja / ko 沿用既有敬体）。术语按 `i18n/GLOSSARY.md`：`Session` 在 zh-CN 作「任务」、ja `セッション`、ko `세션`，`pnpm check:i18n-glossary` 已通过。

（初版此处填的是「不涉及」，不准确——本 PR 确有用户可见的文案新增，已按实际变化改写。）

### 自动验证

```text
pnpm test:unit
结果：全绿，无 FAIL（含 apps/desktop、apps/mobile、packages/maker-core 等全部 workspace）

pnpm --filter @cindy/maker-core exec vitest run src/agents/codex/
结果：12 files / 590 tests passed

pnpm --filter @cindy/maker-core run typecheck
结果：通过

pnpm --filter desktop run typecheck
结果：通过

pnpm check:i18n-glossary
结果：术语表 32 条已裁决 / 35 条待讨论，en / zh-CN / ja / ko 无新增违规

pnpm check:dco
结果：DCO check passed
```

### Review 轮次处理（第 1 轮，已批量收口）

- **check:pr-design-basis（硬门禁）**：初版「引用的设计规范」填「不涉及」——不准确，本 PR 确有用户可见的文案新增；已按实际变化改为引用 DESIGN.md §11.1 / §11.3（见上方「UI 变化」）。改完用 `node scripts/check-pr-design-basis.mjs` 以 PR 真实变更文件列表在本地验证为 PASS 后才推的，不是重跑碰运气。
- **Copilot「无效 usage 后仍挂着 awaitingUsage」→ 修**。确是我引入的缺陷：压缩边界后第一条 usage 若数字不可用（0 / NaN / 缺字段），旧写法先 return、flag 留 true，于是紧随其后的普通 usage（turn 中途累积的水位）会冒名顶替成「压缩后水位」——既可能凭空熔断，也可能把真正的压后水位挤掉而漏判。改为**先消费 flag 再校验取值**，并补了清单要求的完整回归序列。
- **Greptile「last 是增量而非绝对水位」→ 答（已核实证伪，并补测试固定语义）**。查 rollout `019fcd52` 真实 payload 得到两条决定性证据：① `total[n] − total[n−1] === last[n]` 精确成立 → `total` 是 `last` 的累加，`last` 是**单次 API 请求**的量；② `cached[n] ≈ last[n−1]`（28288 vs 28610 / 39552 vs 39722 / 42624 vs 42895 / 76416 vs 76963）→ 本次 prompt 的前一大段命中上次请求建立的 cache，这**只有**在 `last` 是完整 prompt 量时才成立；若它只是「比上次多出来的部分」，cached 不可能等于上一条的完整 input。故 protocol.ts 注释里的「增量」指相对 `total` 累计的本次增量，不是「上下文水位的增量」；单次 LLM 请求的 input 就是该次的完整上下文 = 判据要的绝对水位。证据写进 `noteUsage` 注释，并用**真实 payload**（`total` 与 `last` 不同量级，不再人为设成相同）新增用例固定：同一批数据下 `last` 口径不误杀、`total` 口径会误熔断。
- **Copilot 英文兜底文案语法 → 修**。改为 `over ${elapsedS}s, each leaving about … behind`，并把断言从零散关键词改成整句匹配，防止语法再退化没人发现。
- **Codex「A→B→A 记成 {from:A,to:A}」→ 修**。切回原模型时窗口失配的诱因本就消失了，再提示「切回 A」是让用户重做刚做过的事，文案还自相矛盾；改为**切回 from 即清空记录**，落到不点名原因的兜底文案（那时确实不知道原因，不猜比猜错强）。A→B→C 仍保留最初来源 A 与最新目标 C。两条路径都有回归测试，且 A→B→C 断言到具体模型名，反证 setModel 确实生效、A→B→A 不是假通过。

第 1、2 条属同一族（压缩边界事件消费 + usage 字段语义），已一次覆盖状态消费、无效输入、字段映射与真实 fixture，未拆轮。

新增回归测试（两侧语义都固定住）：
- `compaction-storm.test.ts`（12 例）：直接拿 rollout 实测的 8 个压后水位当 fixture —— 避免判据被改成「合成数据上能过、真实故障上不触发」。覆盖：实测风暴序列熔断、正常压缩不误杀、一次有效压缩即清零、水位不降反涨算无效、非压缩边界的 usage 不参与判定、跨 turn 累计、reset 后重新计数、**无效 usage 消费掉边界后普通 usage 不被顶替**、**last/total 字段语义（真实 payload 固定）**。
- `index.test.ts`：切模型后上游仍报旧窗口 → 按新模型的已核实窗口核算；连续无效压缩 → 终态 error + `turn/interrupt`；正常压缩不熔断；Maker Memory 开启时熔断照常（并断言 debug 日志确证真的走进了 memory 分支，防止该用例退化成副本）；**A→B→A 不再点名模型、A→B→C 保留最初来源与最新目标**。

**已改动一条既有测试的期望**（有意的语义变更，非弱化）：`上报值本来就更小时取上报值(路由真被降窗)` → `上报值更小时仍取已核实窗口`。理由见下方风险。

### 手工验证

不涉及 —— 触发条件是「上游按陈旧窗口反复压缩」，需要真实切模型 + 特定上下文规模才能复现，无法在本地稳定重放。

### 未执行的验证

- **真实切模型场景的实机复现未做**：故障依赖上游 codex-rs 的陈旧窗口行为，本地无法稳定构造。熔断逻辑靠实测数据 fixture 的单测覆盖，实机行为待线上观察 `codex compaction storm escalated to terminal error` 这条 error 日志。
- **UI 双模式目检未做**：本 PR 不新增任何 UI 组件，只增加一条错误文案，走既有 ErrorBanner / ErrorMessageCard 的 themed 渲染路径 —— 但按仓库规则如实声明：两种模式均未实机目检。

### maker-core 核心指标评估（AGENTS.md → maker-core-and-agent-behavior.md §3.4）

改动落在 **usage 计量** 与 **translator/event loop** 路径上，逐项说明：

- **缓存率**：无影响。未触碰 system prompt 任何一段、未改拼接顺序、未在会话中途增删 tool / MCP 注册。`capContextWindow` 只影响本地 usage 快照的分母，不进请求前缀。
- **性能 / 返回速度**：热路径新增开销是 `tokenUsageUpdated` 里一次 `noteUsage()` —— 几个数值比较 + 两个字段赋值，无 IO、无分配、无正则；`handleCompactBoundary` 只在 contextCompaction item（低频）上跑。原先三处 `...(memoryFlushController ? {...} : {})` 每次 item 通知都要建临时对象，改成固定引用后反而少了每事件的对象分配。
- **返回内容准确性**：translator 的事件映射未改，仍是同一个 `compact_boundary` 事件，只是回调改为无条件注册（此前 Maker Memory 关闭时该回调整个不挂，属于漏事件）。model 路由未动，无裸别名。
- **实测结论**：codex 全套 586 测试通过，其中既有的 usage / contextWindow / translator 用例（含缓存与归属语义）全部原样通过，未做任何阈值放宽。

### 风险分类

- [x] 无已知风险
- [ ] SQLite / migration
- [ ] system prompt
- [ ] 协议兼容
- [ ] 权限 / 安全 / 用户数据
- [ ] 存量插件兼容
- [ ] 原生层 / fingerprint / OTA
- [ ] 跨平台差异
- [ ] 其他：

### 影响与回滚

**影响范围**：仅 codex agent 会话。两处行为变化，都请 reviewer 重点看：

1. **误熔断的可能**。判据是「连续 3 次压缩后水位相对上次压缩后下降不足 10%」。正常压缩实测降到 12.7%（241972 → 30738），与风暴期的 <0.5% 差了一个数量级，阈值不需要精调。仍会命中的边缘场景：单条超大用户输入（如贴了 300k 日志）导致压缩后水位下不去 —— 此时压缩确实救不了，熔断并告知用户是正确行为，不是误杀。计数在**任何一次有效压缩**或**用户发新消息**时清零，所以熔断至多让用户多发一条消息，不会锁死会话。

2. **`capContextWindow` 不再采纳更小的上报值**。代价是上游「路由真被降窗」而报出更小值的情形不再被采纳 —— 该场景与「上报值陈旧」在协议上**无法区分**（两者都只是一个更小的数字）。二选一时选了信我们自己按 `(providerId, modelId)` 核实过的那份：它至少不会指向一个用户已经切走的模型。仅在 host 注入了 `resolveVerifiedContextWindow` **且**该路由标了 `contextWindowVerified` 时生效；拿不到已核实窗口时仍沿用上报值（改动前行为）。

**回滚 / 降级方式**：整个 PR 可直接 revert，无数据迁移、无持久化状态、无协议变更。也可单独降级：把 `COMPACTION_STORM_MAX_INEFFECTIVE` 调大即可实际关闭熔断；`capContextWindow` 恢复 `Math.min` 一行即可回到旧语义。

### 上游责任边界（codex-rs，本仓不修）

根因是 codex 在 `thread/settings/update` 切模型后不刷新 `model_context_window`，导致压缩阈值判定用旧模型窗口。证据：rollout `019fcd52` 全程 85 个 `token_count`、11 个 `turn_context` 的该字段恒为 258400；切换后 `turn_context` 里的 `comp_hash: "3000"` 字段也一并消失。网关侧 Opus 5 吃得下 326k、请求全部成功，纯粹是 codex 本地判定超窗。此部分材料已备好，另行向上游提 issue。

### 提交前检查

- [x] 已 review 完整 diff
- [x] 每个 commit 都带 DCO 签名（`git commit -s`，见 [DCO](../DCO)）
- [x] UI 改动已在「UI 变化」注明（本 PR 不涉及）
- [x] 未提交凭证、令牌或授权文件
- [x] 已补充必要文档（判据依据与实测数据写在 `compaction-storm.ts` 头部注释与 `capContextWindow` 注释里）
- [x] 已确认测试结果或说明未执行原因

### CODE REVIEW

**Comment by MagicLizi** [comment]
@zqchris 👋 这个 PR 还有 4 条 review conversation 没 resolve(packages/maker-core/src/agents/codex/compaction-storm.ts / packages/maker-core/src/agents/codex/index.ts),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: packages/maker-core/src/agents/codex/compaction-storm.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+   * 的 usage 反映的是压缩后又累积上去的内容, 拿它比较会把正常增长误判成"没压动"。
+   *
+   * 返回 null 表示这条 usage 不参与判定。
+   */
+  noteUsage(contextTokens: number, now: number): CompactionStormDecision | null {
+    if (!this.awaitingUsage) return null;
+    if (!Number.isFinite(contextTokens) || contextTokens <= 0) return null;
+    this.awaitingUsage = false;
+
+    const previous = this.lastPostCompactionTokens;
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/codex/compaction-storm.ts]: 已修（2ca3916d）。确是本 PR 引入的缺陷：压缩边界后的第一条 usage 若数字不可用（0 / NaN / 缺字段），旧写法先 `return`、`awaitingUsage` 留 `true`，于是紧随其后的普通 usage（turn 中途累积上去的水位）会冒名顶替成「压缩后水位」——既可能凭空熔断，也可能把真正的压后水位挤掉而漏判。

已按你的建议改成**先消费 flag 再校验取值**，并补了完整回归序列：压缩边界 → 无效 usage（消费掉这次机会）→ 两条普通 usage（必须被完全忽略、不建立基线）→ 下一次真正的压缩边界与有效 usage（才重新开始判定，且比较对象是上一次压缩后的水位而非中途那些）。

File: packages/maker-core/src/agents/codex/compaction-storm.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  contextTokens: number;
+  elapsedMs: number;
+  switchedModel?: { from: string; to: string } | null;
+}): string {
+  const elapsedS = Math.round(opts.elapsedMs / 1000);
+  const head =
+    `Codex kept compacting the context without making progress — ` +
+    `${opts.ineffectiveCount} compactions over ${elapsedS}s each left about ` +
+    `${opts.contextTokens} input tokens, so compaction cannot recover this turn. ` +
+    `It was interrupted automatically to stop the loop.`;
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/codex/compaction-storm.ts:225]: 已修（2ca3916d）。改成 `over ${elapsedS}s, each leaving about ${tokens} input tokens behind, so compaction cannot recover this turn.`

这条文案是非 renderer 消费方（IM / orca / 日志）唯一能看到的说明，所以顺手把测试断言从零散关键词改成**整句匹配**，免得语法以后再退化没人发现。

File: packages/maker-core/src/agents/codex/index.ts [hunk 3]
```diff
@@ -7397,6 +7433,41 @@ export class CodexAgent extends BaseAgent {
           const snap = usageTracker.snapshot();
           memoryFlushController.onUsageUpdate(snap.contextTokens, snap.contextWindow);
         }
+        // 压缩风暴熔断 (见 compaction-storm.ts)。喂**上报的 input 总量**而不是
+        // usageTracker 快照: 判据比的是「压缩前后同一口径的水位」, 而 tracker 的
+        // contextTokens 掺了本地口径的换算与跨 turn 累计, 拿它比会把口径差当成压缩效果。
+        const stormDecision = compactionStormTracker.noteUsage(totalInput, Date.now());
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/codex/index.ts:7458]: 核实过了：这条**证伪**，`last.inputTokens` 确实是压缩后的绝对上下文水位，判据取它是对的。但你指出的「现有 fixture 把 total 与 last 人为设成同一水位，无法固定真实映射」完全成立，已用真实 payload 补测试（2ca3916d）。

查 rollout `019fcd52`（本 PR 描述里那次真实故障）的原始 payload，两条决定性证据：

1. **`total[n] − total[n−1] === last[n]` 精确成立** —— total 是 last 的累加，所以 `last` 是**单次 API 请求**的量，不是「上下文水位的增量」：

   | last | cached | total | total 差值 |
   |---|---|---|---|
   | 28610 | 3712 | 28610 | — |
   | 39722 | 28288 | 68332 | 39722 |
   | 42895 | 39552 | 111227 | 42895 |
   | 54840 | 42624 | 166067 | 54840 |

2. **`cached[n] ≈ last[n−1]`**（28288 vs 28610、39552 vs 39722、42624 vs 42895、76416 vs 76963）—— 本次 prompt 的前一大段命中上次请求建立的 prompt cache。这**只有**在 `last.inputTokens` 是本次请求的**完整 prompt token 数**时才可能；若它只是「比上次多出来的部分」，cached 不可能等于上一条的完整 input。

所以 `protocol.ts` 注释里的「上次 turn 的增量」指的是**相对 `total` 累计的本次增量**（即单次请求量），不是上下文水位的增量。而单次 LLM 请求的 input_tokens 就是那次请求的完整上下文 —— 正是判据要比的绝对水位。

已落地：
- 证据链写进 `noteUsage` 的注释，并写明「传 `total` 会拿到单调累加的天文数字，判据直接失效」。
- 新增用例用**真实相邻样本**（total 与 last 不同量级，不再人为设成相同）断言上面两条恒等式/近似关系，并证明同一批数据下 **`last` 口径不误杀、`total` 口径会误熔断** —— 即这个字段选择本身是有判别力的，不是碰巧能过。

File: packages/maker-core/src/agents/codex/index.ts [hunk 4]
```diff
@@ -8852,6 +8920,10 @@ export class CodexAgent extends BaseAgent {
         const prevModel = mutableModel;
         const prevCatalogModel = mutableCatalogModel;
         log.debug('setModel', { from: mutableModel, to: newModel, providerId: mutableProviderId ?? null });
+        // 压缩风暴诊断用 (见 modelSwitchRecord): 首个 from 是 codex 一直拿来算窗口
+        // 的那个模型, 后续切换只更新 to。
+        const prevSwitchRecord = modelSwitchRecord;
+        modelSwitchRecord = { from: modelSwitchRecord?.from ?? prevModel, to: newModel };
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/codex/index.ts]: 已修（2ca3916d）。你说得对，A→B→A 会写出 `{from:A, to:A}`，文案变成「本任务从 A 切到 A，请切回 A」——既自相矛盾，也丢了实际切换目标。

产品语义上选择了**切回 from 时整条清空记录**，而不是保留最后一个不同的目标：切回原模型后，当前模型与 codex 拿来算窗口的模型又一致了，**窗口失配这个诱因本身已经消失**，此时再提示「切回 A」是让用户去做他刚做过的事。清空后若仍压不动，落到不点名原因的兜底文案（`system prompt and tool definitions` 那条）——那时确实不知道原因，不猜比猜错强。

A→B→C 仍按原设计保留最初来源与最新目标（`{from:A, to:C}`）：codex 一直按最初的 A 算窗口，提示要指向 A 而不是中间那跳 B。

两条路径都有回归测试；A→B→C 那条断言到具体模型名（`switched model from "model-a" to "model-c"`、不含 `model-b`），反证 setModel 确实生效、A→B→A 的用例不是假通过。

**Comment by MagicLizi** [comment]
@zqchris 👋 这个 PR 还有 1 条 review conversation 没 resolve(packages/maker-core/src/agents/codex/compaction-storm.ts),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: packages/maker-core/src/agents/codex/compaction-storm.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    // 的水位)被当成压缩后水位, 既可能凭空熔断, 也可能把真正的压后水位挤掉而漏判。
+    this.awaitingUsage = false;
+    if (!Number.isFinite(contextTokens) || contextTokens <= 0) return null;
+
+    const previous = this.lastPostCompactionTokens;
+    this.lastPostCompactionTokens = contextTokens;
+    // 第一次压缩没有可比对象 —— 只记基线, 不判定。
+    if (previous === null) return null;
+
+    const effective = contextTokens < previous * COMPACTION_STORM_EFFECTIVE_DROP_RATIO;
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/codex/compaction-storm.ts]: 已修（63a274d1）。这条成立，而且顺着它复核真实时序时又揪出一个更严重的问题，一并修了。

**1. 判据改为比较同一次压缩的 pre → post**

你说得对：不同 compaction 的 post 之间本就不该被要求单调下降。一个健康的长 autonomous turn 每次压完都回到相近 floor（30k / 31k / 29k / 32k）——那正是压缩**正常工作**的样子，却会被原判据连判无效，几轮后中断一个健康 turn。

改成 `pre = 压缩边界前最后一条有效 usage`。实测那正是 codex 为生成摘要发起的那次调用，输入即当时的完整历史，所以 pre→post 的物理意义很干净：**把 pre 那么大的历史压成摘要之后，总量变成了多少**。

实测分离度（rollout `019fcd52`）：

| 场景 | pre | post | post/pre |
|---|---|---|---|
| 正常压缩 | 241972 | 30738 | **12.7%** |
| 风暴第 2 轮 | 32283 | 325868 | **1009%** |
| 风暴第 3 轮 | 30544 | 325922 | **1067%** |

风暴时 pre 只剩 30k（历史已经压到底了），post 却仍是 326k——膨胀全在 system prompt 与 MCP 工具定义里，压缩碰不到那部分。健康长 turn 的 200k → 30k 一直判有效，不再误杀。

**2. 顺带修掉一个会让熔断彻底失效的 regression（我上一轮自己引入的）**

复核真实事件序列时发现：每个 `contextCompaction` 之后**紧跟一条 `last.inputTokens = 0`**（codex 的压缩完成标记，15:28 那轮连发两条），之后才是压缩后第一个正常请求的水位：

```
usage(pre) → contextCompaction → usage(0)[×1~2] → usage(post)
```

而我上一轮按另一条 review 意见改成了「先消费边界再校验取值」——那条 0 会吃掉边界，真正的 post 被忽略，**熔断在真实时序下永远不会触发**。上一轮的 fixture 里没有那条 0，所以没测出来。

现在改为：跳过非正值继续找 post，同时保留有界（最多 5 条）以免标志悬挂——原意见担心的「flag 悬挂导致后续普通 usage 冒名顶替」和真实时序的需要，两个诉求同时满足。

**3. 测试**

全部按真实事件序列重做（含那些 0），新增四条：长 turn 稳定 floor 不熔断、0 不被当成压后水位、连发两条 0、非正值超上限后放弃边界；集成测试的 `total` 不再与 `last` 设成同值。旧窗口失配序列的熔断覆盖保留（用实测 pre/post 对）。

判别力用两次**变异测试**确证，不是靠“测试通过”自证：
- 判据换回旧版（post vs 上一个 post）→「稳定 floor 不熔断」变红；
- 换回「0 消费边界」→ 16 条里 7 条变红。

codex 套件 594 passed；`pnpm test:unit` 全绿；maker-core / desktop typecheck 通过。

**Comment by MagicLizi** [comment]
@zqchris 👋 这个 PR 还有 1 条 review conversation 没 resolve(apps/desktop/src/renderer/i18n/locales/en/common.json),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: apps/desktop/src/renderer/i18n/locales/en/common.json [hunk 6]
```diff
@@ -8285,6 +8285,7 @@
       "permissionTightenInterruptFailed": "Failed to stop the running task while tightening permissions; it may keep running without approval prompts until it finishes. Stop it manually if needed.",
       "codexAutoReviewUnavailable": "Auto-review is temporarily unavailable. The current action was blocked and permissions switched to Default. Retry the request; Cindy will ask when needed.",
       "upstreamResponseIdleTimeout": "The model stopped responding for a long time, so this turn was interrupted to avoid hanging forever. Completed tool calls are preserved — send the next message to continue.",
+      "codexCompactionNotConverging": "Codex kept compacting the context without making progress, so this turn was interrupted to stop the loop. This usually happens after switching model mid-task — Codex still evaluates the context limit against the previous model. Switch back to that model to continue, or start a new task.",
```
[the remark this answers is not in the document]
Comment (zqchris) [reply apps/desktop/src/renderer/i18n/locales/en/common.json]: 已修（b3519492）。这条完全成立，而且根因比"文案写得太满"更深一层。

**根因：renderer 的 reason 映射会完全覆盖动态 message。**

`ErrorBanner` 与 `ErrorMessageCard` 都是同一句：

```ts
displayError = errorReasonI18nKey ? t(errorReasonI18nKey) : error;
```

所以我上一轮特意写的「没有切换证据就不猜原因」那个兜底 message，在 UI 上**根本看不到**——用户读到的永远是 reason 对应的那条本地化文案。只要两种情形共用一个 reason，那条文案就只有两种下场：无条件断言切模型（对没切过、或 A→B→A 已切回的用户就是无证据误导，而且「切回原模型」无从执行），或者退回不痛不痒的通用话术（浪费我们确实掌握的切换证据）。**只改 message 解决不了这个问题**，这也是我上一轮的疏漏。

**改法：按证据拆两条 reason**

- `codex_compaction_not_converging` —— 通用，不猜原因
- `codex_compaction_not_converging_model_switch` —— 有切换记录才用

默认那条**刻意**留给通用语义：以后万一漏了映射或走到没预料的路径，落到不猜原因的文案是安全的，落到断言切模型的文案是有害的。

同时把 `buildCompactionStormMessage` 改成 `buildCompactionStormTerminalError`，**reason 与 message 一起产出**。两者是同一件事的两种呈现，分开算迟早出现「reason 说因为切模型、message 说不知道原因」的分叉——而因为 renderer 用 reason 盖 message，分叉后错的那半恰好是用户唯一看得到的那半。现在结构上不可能分叉。

四语通用文案一并改写，去掉切模型断言，保留可执行的下一步（新开一个任务），符合 DESIGN.md §11.1「Errors = what happened + what to do」。

**回归测试（就是你要求的那条"钉住 reason 映射不会覆盖无证据的动态 fallback"）**

新增 `apps/desktop/src/renderer/__tests__/compactionStormErrorCopy.test.ts`：
- 两条 reason 各有独立 i18n key（共用一个 key 等于没拆）；
- **通用文案在 en / zh-CN / ja / ko 四语下都不得出现切模型措辞**（各语言分别列了三个断言词）；
- 两条文案都要给出可执行的下一步。

maker-core 侧补：reason/message 同源断言（遍历有无 switchedModel，确认 reason 声称切换 ⟺ message 声称切换 ⟺ 确实有记录）、A→B→A 用通用 reason、A→B→C 用专用 reason。

判别力用**变异测试**确证：把切模型断言塞回通用文案 → 用例立刻变红。

codex 套件 596 passed；desktop 新测试 13 passed；`pnpm test:unit` 全绿；maker-core / desktop typecheck、`check:i18n-glossary` 均通过。

**Comment by MagicLizi** [comment]
@zqchris 👋 这个 PR 还有 1 条 review conversation 没 resolve(packages/maker-core/src/agents/codex/index.ts),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: packages/maker-core/src/agents/codex/index.ts [hunk 7]
```diff
@@ -8852,6 +8932,12 @@ export class CodexAgent extends BaseAgent {
         const prevModel = mutableModel;
         const prevCatalogModel = mutableCatalogModel;
         log.debug('setModel', { from: mutableModel, to: newModel, providerId: mutableProviderId ?? null });
+        // 压缩风暴诊断用 (见 modelSwitchRecord): 首个 from 是 codex 一直拿来算窗口
+        // 的那个模型, 后续切换只更新 to; 切回 from 则整条清空 (诱因已消失)。
+        const prevSwitchRecord = modelSwitchRecord;
+        const switchOrigin = modelSwitchRecord?.from ?? prevModel;
+        modelSwitchRecord =
+          newModel === switchOrigin ? null : { from: switchOrigin, to: newModel };
```
[the remark this answers is not in the document]
Comment (zqchris) [reply packages/maker-core/src/agents/codex/index.ts]: 已修（16ec588f）。核实成立，而且仓内就有代码级证据——`threadSettingsUpdated` 那处的注释写得很明白：

```ts
// 只对齐 wire 值: server 会把请求 id 规范化(`gpt-5.4` → `gpt-5.4-codex`),那个变体
// 目录里没有。**刻意不动 mutableCatalogModel** —— 否则窗口上限会因为查不到目录条目
// 而停止收敛(见该变量注释)。
if (typeof s.model === "string" && s.model) mutableModel = s.model;
```

所以 `mutableModel` 是 wire 口径、`mutableCatalogModel` 才是用户能在选择器里点到的目录 id，而 `setModel` 收到的 `newModel` 来自选择器 = 目录 id。我此前拿 `prevModel` 当切换来源，两种口径混着比，**同时**坏掉这条记录的两个用途：

1. **清除判据失效**：A→B→A 时 `"model-a" !== "model-a-codex"`，记录清不掉，熔断文案继续声称「本任务中途切换过模型」，可用户其实已经切回来了；
2. **诊断指向用户点不到的 id**：文案会让人「切回 model-a-codex」，而选择器是按目录渲染的，那个变体根本不在里面——比不给建议更糟。

改为以 `prevCatalogModel` 为基准；拿不到目录口径来源时（catalog 快照缺失）不记录，走不猜原因的兜底文案。

**关于「一次覆盖该逻辑族的全部入口」**

清点过了：`modelSwitchRecord` 全仓只有 4 处引用——声明、熔断时读取（message + 日志）、`setModel` 写入、`setModel` catch 回滚。**唯一的写入点就是 `setModel`**，所以改这一处即覆盖整族，不存在需要下一轮再补的别处入口。

另外三个会动模型状态的地方都**不该**记录，也确实没有写入点，一并说明免得后来人以为是漏了：
- `thread/start` 的 `gpt-5` 哨兵解析——那是把占位符解析成真实 id，不是用户切换；
- `thread/resume` 的 hydrate——恢复会话，不是用户切换；
- `threadSettingsUpdated` 的 wire 对齐——server 规范化，不是用户切换。

判据与理由已写进 `modelSwitchRecord` 的注释，免得以后有人顺手换回 `mutableModel`。

**回归测试**

新增两条，都模拟 server 规范化（推一条 `thread/settings/updated` 把 `mutableModel` 变成 `model-a-codex`，`mutableCatalogModel` 按设计保持 `model-a`）：
- A→B→A 仍能清除记录 → 通用 reason，且断言 `model-a-codex` **绝不出现**在文案里；
- A→B→C 的诊断仍用目录 id（`from "model-a" to "model-c"`），同样断言不含 wire 变体。

判别力用变异测试确证：基准换回 `prevModel` → 这两条立刻变红（也就证明了这个 bug 在真实 wire 规范化下确实会发生，不是理论推演）。

codex 套件 598 passed；`pnpm test:unit` 全绿；maker-core / desktop typecheck、`check:i18n-glossary` 均通过。

**Review by MagicLizi — APPROVED** [review]
Code review passed — compaction storm detection is well-implemented with thorough real-data-driven tests, clean edge case handling (zero-value post-compaction markers, A→B→A switch tracking, wire ID normalization), and reasonable tradeoff on capContextWindow. The i18n messages correctly split generic vs model-switch diagnostics per evidence.

**Comment by MagicLizi** [comment]
实测数据定位到 codex 窗口失配的根因后，熔断器的判据精准地只看「压了等于没压」这个稳定特征——不碰窗口值本身，把不可控的上游状态降格成可收敛的本地断言。测试里连那条真实的零值标记都没放过，扎实。
